urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/RangeCoder/BitTreeEncoder.java

```diff
@@ -0,0 +1,102 @@
+package SevenZip.Compression.RangeCoder;
+
+import java.io.IOException;
+
+public class BitTreeEncoder
+{
+    short[] Models;
+    int     NumBitLevels;
+    
+    public BitTreeEncoder(int numBitLevels)
+    {
+        NumBitLevels = numBitLevels;
+        Models = new short[1 << numBitLevels];
+    }
+    
+    public void Init()
+    {
+        Decoder.InitBitModels(Models);
+    }
+    
+    public void Encode(Encoder rangeEncoder, int symbol) throws IOException
+    {
+        int m = 1;
+        for (int bitIndex = NumBitLevels; bitIndex != 0;)
+        {
+            bitIndex--;
+            final int bit = symbol >>> bitIndex & 1;
+            rangeEncoder.Encode(Models, m, bit);
+            m = m << 1 | bit;
+        }
+    }
+    
+    public void ReverseEncode(Encoder rangeEncoder, int symbol)
+            throws IOException
+    {
+        int m = 1;
+        for (int i = 0; i < NumBitLevels; i++)
+        {
+            final int bit = symbol & 1;
+            rangeEncoder.Encode(Models, m, bit);
+            m = m << 1 | bit;
+            symbol >>= 1;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 37:43, 94:100.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.